libs/header/p2p/exchange.go

```diff
@@ -226,22 +226,30 @@ func (ex *Exchange[H]) performRequest(
 	if len(ex.trustedPeers) == 0 {
 		return nil, fmt.Errorf("no trusted peers")
 	}
-
-	for {
+	var reqErr error
+	for i := 0; i < len(ex.trustedPeers); i++ {
 		//nolint:gosec // G404: Use of weak random number generator
 		idx := rand.Intn(len(ex.trustedPeers))
-		ctx, cancel := context.WithTimeout(ctx, ex.Params.TrustedPeersRequestTimeout)
 
-		h, err := ex.request(ctx, ex.trustedPeers[idx], req)
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		reqCtx, cancel := context.WithTimeout(ctx, ex.Params.TrustedPeersRequestTimeout)
+		h, err := ex.request(reqCtx, ex.trustedPeers[idx], req)
 		cancel()
 		switch err {
 		default:
+			reqErr = err
 			log.Debug(err)
 			continue
-		case context.Canceled, context.DeadlineExceeded, nil:
-			return h, err
```

Comment:
Why?

Reply:
in case we perform `len(ex.trustedPeers)` that will fail , we will return the last stored error. If `context.DeadlineExceeded`, it's fine IMO and we can try to get header from another peer.

Reply:
reworked